File: 0201-0300/0204.Count_Primes.py
#4748ms



class Solution:
    def countPrimes(self, n: int) -> int:
        # If n <= 2, there are no primes less than n
        if n <= 2:
            return 0

        # Boolean array where is_prime[i] tells whether i is prime
        # Initially assume all numbers are prime
        is_prime = [True] * n

        # 0 and 1 are not prime
        is_prime[0] = False
        is_prime[1] = False

        # List to store discovered prime numbers
        primes = []

        # Iterate through all numbers from 2 to n-1
        for i in range(2, n):

            # If still marked prime, it is a prime number
            if is_prime[i]:
                primes.append(i)

            # Mark multiples of i using previously found primes
            # This prevents checking divisibility repeatedly
            for p in primes:

                # Stop if product goes out of range
                if i * p >= n:
                    break

                # Mark composite number
                is_prime[i * p] = False

        # Number of primes found
        return len(primes)




#0ms




class Solution:
    def countPrimes(self, n: int) -> int:
        if not n:
            return 0
        return cumulative[n - 1]

N = 5 * pow(10, 6)
is_prime = [1] * N
is_prime[0] = is_prime[1] = 0

for n in range(4, N, 2):
    is_prime[n] = 0

for num in range(3, floor(sqrt(N)) + 1, 2):
    if is_prime[num]:
        for multiple in range(num * num, N, num):
            is_prime[multiple] = 0
cumulative = list(accumulate(is_prime))





#73ms



class Solution:
    def countPrimes(self, n: int) -> int:
        if n <= 2:
            return 0
        
        # Initialize bytearray for odd numbers less than n.
        # The size is n // 2.
        # Index i represents the number (2 * i + 1).
        sieve = bytearray([1]) * (n // 2)
        
        # Index 0 represents number 1, which is not prime.
        sieve[0] = 0
        
        # We only need to iterate up to the square root of n.
        for i in range(3, int(n**0.5) + 1, 2):
            # Check if index corresponds to a prime number
            # (i - 1) // 2 converts number 'i' to its corresponding index
            if sieve[(i - 1) // 2]:
                # If i is prime, mark all its odd multiples as non-prime.
                # The first odd multiple to mark is i * i.
                # The step is 2 * i (since odd + odd = even, we skip even multiples).
                # In our index space (0, 1, 2...), a step of 2*i in numbers corresponds to a step of i in indices.
                start_index = (i * i - 1) // 2
                
                # We update the slice using standard slice assignment.
                # The length calculation ensures we generate a bytearray of the exact needed size.
                sieve[start_index::i] = bytearray((len(sieve) - 1 - start_index) // i + 1)
        
        # sum(sieve) counts the odd primes. We add 1 to include the even prime '2'.
        return sum(sieve) + 1

File: 0201-0300/0204.Count_Primes.cpp
//39ms



class Solution {
public:
    int countPrimes(int n) {
        if (n < 3) {
            return 0;
        }
        bool isprime[n];

        memset(isprime, true, n);

        int result = n / 2;
        for (int i = 3; i * i < n; i += 2) {
            if (isprime[i]) {
                int d = i * 2;
                for (int j = i * i; j < n; j += d) {
                    if (isprime[j]) {
                        isprime[j] = false;
                        result--;
                    }
                }
            }
        }
        return result;
    }
};





//0ms





int MAX = 6e6;

vector<bool> GenerateSieve() {
    vector<bool> isprime(MAX, true);

    isprime[0] = false;
    isprime[1] = false;

    for (int i = 2; i * i < MAX; i++){
        if (isprime[i]) {
            for(int j = i * i; j < MAX; j += i) {
                if (isprime[j]) {
                    isprime[j] = false;
                }
            }
        }
    }

    return isprime;
}

vector<bool> isprime = move(GenerateSieve());

vector<int> GenerateAns() {
    vector<int> ans(MAX, 0);

    ans[0] = 0;
    ans[1] = 0;

    for (int i = 2; i < MAX; i++)
        ans[i] = ans[i - 1] + isprime[i - 1];

    return ans;
}

vector<int> ans = move(GenerateAns());

class Solution {
public:
    int countPrimes(int n) {
        return ans[n];
    }
};





//198ms







class Solution {
public:
    int countPrimes(int n) 
    {
        vector<bool> primes(n+1,true);
        primes[0]=false;
        primes[1]=false;
        for(long long i=2;i<=n;i++) if(primes[i]) for(long long j=i*i;j<=n;j+=i) primes[j]=false;
        int ans=0;
        for(int i=2;i<n;i++) if(primes[i]) ans++;
        return ans;
    }
};







//18ms





class Solution {
    const static int N = 5000005;
    static bool done;
    static char isPrime[N];   // char is faster than vector<bool>
    static int primeCount[N]; // prefix sum of primes

    void sieve() {
        if (done) return;
        for (int i = 0; i < N; i++) isPrime[i] = 1;
        isPrime[0] = isPrime[1] = 0;

        for (int i = 2; i*i < N; i++) {
            if (isPrime[i]) {
                for (int j = i*i; j < N; j += i)
                    isPrime[j] = 0;
            }
        }

        primeCount[0] = 0;
        for (int i = 1; i < N; i++)
            primeCount[i] = primeCount[i-1] + isPrime[i];

        done = true;
    }

public:
    int countPrimes(int n) {
        sieve();
        if (n < 2) return 0;
        return primeCount[n-1]; // primes strictly less than n
    }
};

// static members
bool Solution::done = false;
char Solution::isPrime[Solution::N];
int Solution::primeCount[Solution::N];

















